File: backend/AwsAccounting.Api/Reconciliation/MappingResolver.cs
using System.Text.Json;
using AwsAccounting.Api.Data;
using AwsAccounting.Api.Domain;
using Microsoft.EntityFrameworkCore;

namespace AwsAccounting.Api.Reconciliation;

public record ResolvedMapping(ColumnMapping Mapping, string Source);

/// <summary>Identifies a column mapping via the configured AI when auto-detect leaves gaps. P3 wires a live provider; until then it declines.</summary>
public interface IAiMappingIdentifier
{
    Task<ColumnMapping?> IdentifyAsync(IReadOnlyList<string[]> rows, ColumnMapping auto, CancellationToken ct);
}

/// <summary>No-op AI identifier used until the AI layer (P3) is wired in.</summary>
public sealed class NullAiMappingIdentifier : IAiMappingIdentifier
{
    public Task<ColumnMapping?> IdentifyAsync(IReadOnlyList<string[]> rows, ColumnMapping auto, CancellationToken ct)
        => Task.FromResult<ColumnMapping?>(null);
}

/// <summary>
/// Resolves a column mapping without asking the user:
///   1. reuse a learned mapping for this exact header layout (per tenant), else
///   2. auto-detect (store it if complete), else
///   3. ask the configured AI to identify the mapping (store it), else
///   4. return the best-effort auto mapping (the run surfaces any gaps).
/// A tenant can have many learned mappings — one per distinct format fingerprint.
/// </summary>
public sealed class MappingResolver(AppDbContext db, IAiMappingIdentifier ai, ILogger<MappingResolver> log)
{
    private static readonly JsonSerializerOptions Json = new(JsonSerializerDefaults.Web);

    public async Task<ResolvedMapping> ResolveAsync(Guid tenantId, IReadOnlyList<string[]> rows, string side, CancellationToken ct = default)
    {
        var fp = Mapper.Fingerprint(rows);

        var learned = await db.LedgerMappings
            .AsNoTracking()
            .FirstOrDefaultAsync(x => x.TenantId == tenantId && x.Fingerprint == fp, ct);
        if (learned != null)
        {
            var m = JsonSerializer.Deserialize<ColumnMapping>(learned.Mapping, Json) ?? Mapper.AutoDetect(rows);
            // header row can vary between files of the same format — recompute it, keep the learned columns.
            m.HeaderRow = Mapper.DetectHeaderRow(rows);
            return new ResolvedMapping(m, "learned");
        }

        var auto = Mapper.AutoDetect(rows);
        if (Mapper.Gaps(auto).Count == 0)
        {
            await StoreAsync(tenantId, fp, side, auto, "auto", ct);
            return new ResolvedMapping(auto, "auto");
        }

        try
        {
            var aiMap = await ai.IdentifyAsync(rows, auto, ct);
            if (aiMap != null && Mapper.Gaps(aiMap).Count == 0)
            {
                await StoreAsync(tenantId, fp, side, aiMap, "ai", ct);
                return new ResolvedMapping(aiMap, "ai");
            }
        }
        catch (Exception e)
        {
            log.LogWarning(e, "AI mapping identification failed for tenant {TenantId}", tenantId);
        }

        return new ResolvedMapping(auto, "auto-partial");
    }

    private async Task StoreAsync(Guid tenantId, string fp, string side, ColumnMapping mapping, string source, CancellationToken ct)
    {
        var json = JsonSerializer.Serialize(mapping, Json);
        var existing = await db.LedgerMappings.FirstOrDefaultAsync(x => x.TenantId == tenantId && x.Fingerprint == fp, ct);
        if (existing == null)
        {
            db.LedgerMappings.Add(new LedgerMapping { TenantId = tenantId, Fingerprint = fp, Side = side, Mapping = json, Source = source });
        }
        else
        {
            existing.Mapping = json;
            existing.Source = source;
            existing.UpdatedAt = DateTimeOffset.UtcNow;
        }
        await db.SaveChangesAsync(ct);
    }
}

File: frontend/src/modules/ar-reconciliation/mapping-resolver.ts
import { createHash } from "node:crypto";
import { and, eq } from "drizzle-orm";
import { db } from "../../db";
import { ledgerMappings } from "../../db/schema";
import { autoDetectMapping, mappingGaps, detectHeaderRow, type ColumnMapping, type CanonicalField } from "./ledger-mapping";
import { aiJson, AiNotConfiguredError } from "../../lib/ai";

/** A stable fingerprint of a ledger's header layout (order-independent). */
export function fingerprint(rows: string[][]): string {
  const headers = (rows[detectHeaderRow(rows)] ?? [])
    .map((h) => (h ?? "").toString().trim().toLowerCase().replace(/[^a-z0-9]+/g, ""))
    .filter(Boolean)
    .sort()
    .join("|");
  return createHash("sha256").update(headers).digest("hex").slice(0, 32);
}

const num = (v: unknown): number => (typeof v === "number" && Number.isInteger(v) ? v : -1);

/**
 * Resolve a column mapping without asking the user:
 *   1. reuse a learned mapping for this exact header layout (per tenant), else
 *   2. auto-detect (store it if complete), else
 *   3. ask the configured AI to identify the mapping (store it), else
 *   4. return the best-effort auto mapping (executeRun surfaces any gaps).
 * A tenant can have many learned mappings — one per distinct format fingerprint.
 */
export async function resolveMapping(
  tenantId: string,
  rows: string[][],
  side: "statement" | "customer",
): Promise<{ mapping: ColumnMapping; source: string }> {
  const fp = fingerprint(rows);

  const [learned] = await db
    .select()
    .from(ledgerMappings)
    .where(and(eq(ledgerMappings.tenantId, tenantId), eq(ledgerMappings.fingerprint, fp)))
    .limit(1);
  if (learned) {
    const m = learned.mapping as ColumnMapping;
    // header row can vary between files of the same format — recompute it, keep the learned columns.
    return { mapping: { ...m, headerRow: detectHeaderRow(rows) }, source: "learned" };
  }

  const auto = autoDetectMapping(rows);
  if (mappingGaps(auto).length === 0) {
    await store(tenantId, fp, side, auto, "auto");
    return { mapping: auto, source: "auto" };
  }

  try {
    const aiMap = await aiIdentifyMapping(rows, auto);
    if (mappingGaps(aiMap).length === 0) {
      await store(tenantId, fp, side, aiMap, "ai");
      return { mapping: aiMap, source: "ai" };
    }
  } catch (e) {
    if (!(e instanceof AiNotConfiguredError)) console.error("AI mapping failed:", e);
  }

  return { mapping: auto, source: "auto-partial" };
}

async function store(tenantId: string, fp: string, side: string, mapping: ColumnMapping, source: string) {
  await db
    .insert(ledgerMappings)
    .values({ tenantId, fingerprint: fp, side, mapping, source })
    .onConflictDoUpdate({
      target: [ledgerMappings.tenantId, ledgerMappings.fingerprint],
      set: { mapping, source, updatedAt: new Date() },
    });
}

async function aiIdentifyMapping(rows: string[][], auto: ColumnMapping): Promise<ColumnMapping> {
  const headerRow = auto.headerRow;
  const headers = rows[headerRow] ?? [];
  const sample = rows.slice(headerRow + 1, headerRow + 6);
  const system = "You map accounting-ledger spreadsheet columns to canonical fields. Respond with JSON only, no prose.";
  const user =
    `Header cells (0-indexed): ${JSON.stringify(headers)}\n` +
    `Sample data rows: ${JSON.stringify(sample)}\n\n` +
    `Return JSON exactly: {"reference":n,"date":n,"description":n,"debit":n,"credit":n,"amount":n,"amountMode":"debit_credit"|"signed","positiveIsDebit":true|false} ` +
    `where n is the 0-based column index or -1 if absent. Use "debit_credit" when there are separate debit and credit columns; use "signed" for a single +/- amount column (then set positiveIsDebit).`;

  const j = await aiJson<Record<string, unknown>>({ purpose: "reasoning", system, user });
  const columns: Record<CanonicalField, number> = {
    reference: num(j.reference),
    date: num(j.date),
    description: num(j.description),
    debit: num(j.debit),
    credit: num(j.credit),
    amount: num(j.amount),
  };
  const amountMode = j.amountMode === "signed" ? "signed" : columns.debit >= 0 || columns.credit >= 0 ? "debit_credit" : "signed";
  return { columns, amountMode, positiveIsDebit: j.positiveIsDebit !== false, headerRow };
}
